### DISCUSSION TITLE
On JWT

### DISCUSSION THREAD

**Discussion by Fidel-C** [piccolo-orm/piccolo_api#222] [General]
When using JWT, Is there a way to exempt certain endpoints from requiring token in the headers?
I'd also like to know if there's a discord channel or something to join the community.

**Comment by sinisaos** [comment]
@Fidel-C Yes if you write your own endpoints because all non-JWTMiddleware endpoints do not require a JWT token in the header. If you look at the [example in the docs](https://piccolo-api.readthedocs.io/en/latest/jwt/example.html), only the endpoints in the `private_app` are protected and all potential endpoints in the `public_app` are not.
If you want to use `FastAPIWrapper`, maybe you can try to make JWT auth as dependencies and then try to use something like [this example](https://piccolo-api.readthedocs.io/en/latest/fastapi/index.html#oauth2-example), but I haven't tried that.

**Reply by dantownsend** [reply]
I was thinking we could add an `allow_unauthenticated` argument to `JWTMiddleware`. We do this with session auth.

It means the application would have to check in each endpoint if the user was in the scope, and if not reject them. This means the entire application can be wrapped in `JWTMiddleware`, and if some endpoints don't need protecting, just don't check for the user in the scope.

I think it's a 5 minute change - I might see quickly.

**Reply by sinisaos** [reply]
@dantownsend That's great idea.

**Reply by Fidel-C** [reply]
@dantownsend that would be great.

**Reply by dantownsend** [reply]
I've created a PR for it https://github.com/piccolo-orm/piccolo_api/pull/224

**Reply by dantownsend** [reply]
I've released the ``allow_unauthenticated`` feature in ``piccolo-api==0.54.0``.

**Reply by Fidel-C** [reply]
> I've released the `allow_unauthenticated` feature in `piccolo-api==0.54.0`.

Wow! good job .

**Comment by Fidel-C** [comment]
@sinisaos might try that out too.  Thanks.

**Comment by dantownsend** [comment]
> I'd also like to know if there's a discord channel or something to join the community.

We don't use Discord much at the moment - GitHub discussions is quite nice because we can easily convert discussions into issues. But we probably should use Discord more as people do like it.

**Comment by Fidel-C** [comment]
I observed that once I secure a particular app either using Session, Jwt, Token, it disappeares completely from the docs. I suppose that this would be difficult for a front-end developer who needs the docs. Adding allow_unauthenticated flag to the jwt authentication reveals the docs but then it also removes authentication.
My expectation was to see the protected routes in the docs but with a locked padlock icon indicating that they require authentication. I stand to be corrected please. Maybe I'm doing something wrong.

**Reply by dantownsend** [reply]
If using session auth, if you're logged into Piccolo Admin, you should be able to access the endpoints (as your session cookie will be sent automatically with any requests).

We have a custom Swagger endpoint which works with Piccolo's CSRF protection, which you may prefer to FastAPI's built in one:

https://piccolo-api.readthedocs.io/en/latest/openapi/index.html

As for token and JWT auth, the work around I use is to install a Chrome extension like [ModHeader](https://chrome.google.com/webstore/detail/modheader-modify-http-hea/idgpnmonknjnojddfkpgkljpfnnfcklj). That lets you add a token header which means the Swagger docs will work correctly.

There may be something better we can do to make the Swagger docs display an input for token headers ([maybe this](https://swagger.io/docs/specification/authentication/bearer-authentication/)).

**Reply by sinisaos** [reply]
@Fidel-C If you use Piccolo Session, JWT, or Token auth, you are protecting the entire app with the middleware, not just some endpoints in the app. If you use FastAPI and FastAPIWrapper, you can try [FastAPI’s builtin OAuth2](https://piccolo-api.readthedocs.io/en/latest/fastapi/index.html#oauth2-example) to protect endpoints with insecure HTTP methods. More examples of that approach is [here](https://github.com/sinisaos/headless-forum-fastapi) and [here](https://github.com/sinisaos/piccolo-vue-starter). I hope you find something useful in these examples.

**Reply by Fidel-C** [reply]
@dantownsend and @sinisaos thanks a lot for your help. Will have to implement the 0Auth2 approach then. Will still checkout ModHeader extension too.

**Reply by dantownsend** [reply]
Looks like you can do this:

```python
from fastapi import Security
from fastapi.security.api_key import APIKeyHeader


auth_header1 = APIKeyHeader(name="Authorization")


@my_router.get("/")
async def my_endpoint(header=Security(auth_header1)):
    ...
```

This then allows you to specify a header token in the Swagger docs:

<img width="680" alt="Screenshot 2023-04-10 at 20 57 53" src="https://user-images.githubusercontent.com/350976/230986545-2c14588b-e1c7-47d7-a627-00c4fbb2ded2.png">

**Reply by dantownsend** [reply]
So there's way of allowing the user to specify a token via the Swagger docs.

The remaining challenge is if you wrap the entire app in say `JWTMiddleware` it also protects the `docs` endpoint which is something we don't want.

One work around for this is we add a new argument to `JWTMiddleware` which lets us ignore certain URLs. For example:

```python
app = JWTMiddleware(my_asgi_app, unprotected_paths=['/docs'])
```

@sinisaos What do you think?

**Reply by Fidel-C** [reply]
This will basically do the magic

**Reply by sinisaos** [reply]
> The remaining challenge is if you wrap the entire app in say `JWTMiddleware` it also protects the `docs` endpoint which is something we don't want.
> 
> One work around for this is we add a new argument to `JWTMiddleware` which lets us ignore certain URLs. For example:
> 
> ```python
> app = JWTMiddleware(my_asgi_app, unprotected_paths=['/docs'])
> ```

@dantownsend But isn't that the same as when set `allow_unauthenticated=True` to `JWTMiddleware` (we can see the docs on `/docs` route, but the endpoints in docs are unprotected and we don't want that)? Piccolo middlewares is really great if we want to protect the entire app, but if we want to protect only individual endpoints in the app, I think the best way is to use [FastAPI built in OAuth](https://github.com/sinisaos/headless-forum-fastapi/blob/master/screenshots/api_docs.png) because that way we can pass auth as dependency to every route we want. After user is logged in (using Authorize button or route lock icon) on all protected endpoints a jwt token will also be sent in Authorization header of request.

![request](https://user-images.githubusercontent.com/30960668/231072613-41c65be1-82ab-4a02-829f-82bca704b7e0.png)

**Reply by Fidel-C** [reply]
@sinisaos setting allow_unauthenticated=True removes the authentication from all routes.

**Reply by sinisaos** [reply]
@Fidel-C Yes and I don't see how it will be any different with this proposed API
```python
app = JWTMiddleware(my_asgi_app, unprotected_paths=['/docs'])
```
If I understood correctly (apologies if not) with this the `/docs` route will be unprotected and the `/docs` route can have multiple tables with multiple endpoints. How can we protect for example `POST` endpoints in one table and not another? With FastAPI built-in OAuth we can do that. If you look at this [image](https://github.com/sinisaos/headless-forum-fastapi/blob/master/screenshots/api_docs.png), you can see that it only protects certain endpoints. I apologize if I missed the point.

**Reply by dantownsend** [reply]
@sinisaos Using FastAPI's OAuth features is an option. In some of my apps I have a blend of Starlette and FastAPI, so using generic ASGI middleware is preferable, though for some people this doesn't matter.

With `JWTMiddleware` the `allow_unauthenticated` option makes the user check in their endpoints that the user is authenticated rather than rejecting them automatically.

Lets say you have 10 API endpoints, and the `/docs/` endpoint. You can add a check to the 10 endpoints to check that the user is authenticated. And that means the `/docs/` endpoint is still accessible.

Alternatively, if we added something like `unprotected_paths=['/docs']` you wouldn't have to add any checks to those 10 endpoints - they would all be automatically protected besides the `/docs/` endpoint.

It's just an idea.

There has been talk in FastAPI for a while about allowing [middleware on routers](https://github.com/tiangolo/fastapi/issues/1174). This would make things much easier, as currently you have to attach middleware at the app level. I usually work around this by having a public FastAPI app, and mounting another FastAPI app to it which is protected with middleware.

**Reply by sinisaos** [reply]
> Alternatively, if we added something like `unprotected_paths=['/docs']` you wouldn't have to add any checks to those 10 endpoints - they would all be automatically protected besides the `/docs/` endpoint.
> 
> It's just an idea.

@dantownsend OK. It would be a different story if the endpoints were also protected in Openapi `/docs` endpoint. It would be something similar to Starlite [guards](https://docs.litestar.dev/latest/usage/security/guards.html#guards) if I understand correctly. I have successfully protected endpoints in Swagger docs in Starlite app [here](https://github.com/sinisaos/starlite-piccolo/blob/main/backend/accounts/guards.py) and [here](https://github.com/sinisaos/starlite-piccolo/blob/main/backend/tasks/endpoints.py). If you make a POC I will be happy to try and test the new features.

**Reply by sinisaos** [reply]
@dantownsend Did you mean something like this?
```python
class JWTMiddleware:

    def __init__(
        ...
        visible_paths: t.List[str] = [],
    ) -> None:
        ...
        self.visible_paths = visible_paths

    async def __call__(self, scope, receive, send):
        """
        Add the user_id to the scope if a JWT token is available, and the user
        is recognised, otherwise raise a 403 HTTP error.
        """
        allow_unauthenticated = self.allow_unauthenticated
        visible_paths = self.visible_paths

        headers = dict(scope["headers"])
        token = self.get_token(headers)
        if not token:
            error = JWTError.token_not_found.value
            if allow_unauthenticated or scope["path"] in visible_paths: # we need to add this line to each allow_unauthenticated check.
                await self.asgi(
                    extend_scope(scope, {"user_id": None, "jwt_error": error}),
                    receive,
                    send,
                )
                return
            else:
                raise HTTPException(status_code=403, detail=error)
           ...
```
With that, we can see the Swagger docs in browser, but we can't interact with the docs because it's token protected.

![docs](https://user-images.githubusercontent.com/30960668/231556240-b3cb0cf4-c709-4fd3-8165-fc8f1967631a.png)

**Reply by dantownsend** [reply]
@sinisaos Yeah, something like that.

> With that, we can see the Swagger docs in browser, but we can't interact with the docs because it's token protected.

You can add make it so the Swagger UI authorisation box lets you specify a token to use as a header.

It's something like this:

```python
from fastapi import Security, Depends, FastAPI
from fastapi.security.api_key import APIKeyHeader


auth_header1 = APIKeyHeader(name="Authorization")

app = FastAPI(dependencies=[Depends(Security(auth_header1))]
```

**Reply by Fidel-C** [reply]
Sorry for the little digression.
I've been trying to implement the normal Fastapi OAuth2
It appears a different hashing scheme was used for piccolo BaseUser password.
So using bcrypt scheme, as in,

from passlib.context import CryptContext
pwd_context=CryptContext(schemes["bcrypt"] , deprecated="auto"]
pwd_context.verify(password,hashed_password)


throws errors showing that it doesn't recognise the hashing scheme used to hash and save the password in the BaseUser table.

**Reply by sinisaos** [reply]
@Fidel-C [Here](https://github.com/sinisaos/headless-forum-fastapi/blob/master/accounts/endpoints.py) is example how to use Piccolo BaseUser with FastAPI built-in OAuth.

@dantownsend I changed your example to this and it works. If you want I can do PR to add `visible_paths`.
```python
from fastapi import Depends, FastAPI
from fastapi.security.api_key import APIKeyHeader

public_app = FastAPI(...)

auth_header = APIKeyHeader(name="Authorization")
private_app = FastAPI(dependencies=[Depends(auth_header)]

protected_app = JWTMiddleware(
    private_app,
    auth_table=BaseUser,
    secret="mysecret123",
    blacklist=MyBlacklist(),
    visible_paths=["/docs", "/openapi.json"],
)


FastAPIWrapper(
    "/bands/",
    fastapi_app=private_app,
    piccolo_crud=PiccoloCRUD(Band, read_only=False),
    fastapi_kwargs=FastAPIKwargs(
        all_routes={"tags": ["Band"]},
    ),
)

public_app.mount("/private", protected_app)
```
Result is:

https://user-images.githubusercontent.com/30960668/231664154-9bd81f2e-b50c-4e13-a37c-f039ff760d4b.mp4

**Reply by dantownsend** [reply]
> It appears a different hashing scheme was used for piccolo BaseUser password.

@Fidel-C `BaseUser` uses `SHA256` for hashing passwords.

https://github.com/piccolo-orm/piccolo/blob/05048ab9d58611795924565965ea6170e265a323/piccolo/apps/user/tables.py#L151

We would have to refactor it to support other hashing algorithms. It's not impossible, but is a bit of work.

**Reply by Fidel-C** [reply]
@sinisaos and @dantownsend thanks a lot for your assistance. Very grateful.

**Reply by sinisaos** [reply]
@Fidel-C I made a PR #226 based on @dantownsend idea, so we'll see what happens. I hope it is good enough because with these changes we will be able to use Swagger docs and authorize the user like in the video I posted.

**Reply by Fidel-C** [reply]
Great work

**Comment by Fidel-C** [comment]
@sinisaos I suppose this FadtApi OAuth2 implementation will work after the update or does it work right away?

**Reply by sinisaos** [reply]
@Fidel-C I'm sorry, but I don't understand. Which FastAPI OAuth implementation, from this [example](https://github.com/sinisaos/headless-forum-fastapi) or from the PR? The code from my example repository will work with or without changes because that code does not use the Piccolo middleware. As for PR, I'm not a maintainer and the code from PR has to pass Daniel's review first and if it's good enough, PR will be merged into master. If you want to try the code from the PR, you must patch your local Piccolo API installation with the code from the PR.

**Reply by Fidel-C** [reply]
@sinisaos thanks a lot. I understand.
